clang/lib/CodeGen/Targets/AMDGPU.cpp

```diff
@@ -52,6 +52,17 @@ class AMDGPUABIInfo final : public DefaultABIInfo {
   void computeInfo(CGFunctionInfo &FI) const override;
   RValue EmitVAArg(CodeGenFunction &CGF, Address VAListAddr, QualType Ty,
                    AggValueSlot Slot) const override;
+
+  llvm::FixedVectorType *
+  getOptimalVectorType(llvm::FixedVectorType *T,
+                       const LangOptions &Opt) const override {
+    // We have legal instructions for 96-bit so 3x32 can be supported.
+    // FIXME: This check should be a subtarget feature as technically SI doesn't
+    // support it.
+    if (T->getNumElements() == 3 && getDataLayout().getTypeSizeInBits(T) == 96)
```

Comment:
@arsenm I wonder if we want to relax it to simply check size to be 96 to support cases like `3x32` (which is covered at the moment), `6x16` and `12x8`.

Reply:
Yes, it should be any 96-bit access

Reply:
It looks like they are preserved anyway. Clang only has special handling of 3-element vector.